### PR TITLE
Add Gradient API

### LINKED ISSUE #928: Expose gradient functionality in API
This should not require the use of MiniMessage.

### PR DESCRIPTION
**Pull request by thiccaxe**
closes #928

### CODE REVIEW

**Comment by thiccaxe** [comment]
- [x] gradient API  
- [ ] update minimessage to use this API
  - [ ] gradients
    - [ ] stops? 
  - [ ] rainbows
- [X] deprecate `TextColor#lerp`
- [X] add other gradients
  - [X] figure out how to deal with the new colorspace 
  - [X] Linear RGB
  - [X] OkLab
  - [ ] OkLCH
- [ ] tests
  - [ ] RGB
  - [ ] HSV
  - [ ] LAB
  - [ ] OKLAB
  - [ ] LCH
  - [ ] OKLCH
  - [ ] minimessage

File: api/src/main/java/net/kyori/adventure/text/format/TextColor.java [hunk 1]
```diff
@@ -261,6 +261,8 @@ public interface TextColor extends Comparable<TextColor>, Examinable, RGBLike, S
    * @param b the upper bound ({@code t=1.0})
    * @return the interpolated value, a color between the two input colors {@code a} and {@code b}
    * @since 4.8.0
+   *
+   * @deprecated see {@link net.kyori.adventure.text.format.gradient.RGBGradient#interpolate(float, TextColor, TextColor)}
```
Comment (MiniDigger) [inline api/src/main/java/net/kyori/adventure/text/format/TextColor.java:265]: Idk if it's just me, but it seems weird to have interpolation methods on the gradient? I don't really have an better API design but something about that feels off to me. Like, they are just util methods and can be used for other stuff, its a bit weird to tie them to the concept of a gradient?
Will take another look when am more awake tho

Comment (thiccaxe) [reply api/src/main/java/net/kyori/adventure/text/format/TextColor.java:265]: Maybe it's poor naming choice. But it doesn't make sense to have an interpolation method on TextColor when the exact same method exists in a more comprehensive API

Comment (kashike) [reply api/src/main/java/net/kyori/adventure/text/format/TextColor.java:265]: The method should stay here on `TextColor`.

File: api/src/main/java/net/kyori/adventure/text/format/gradient/LinearRGBGradient.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+import net.kyori.adventure.text.format.TextColor;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A gradient in the Linear RGB Colorspace.
+ *
+ * @since 4.15.0
+ */
+public class LinearRGBGradient implements Gradient {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/text/format/gradient/LinearRGBGradient.java:34]: All API types should be `interface` - implementations should be package-private.

File: api/src/main/java/net/kyori/adventure/text/format/gradient/HSVGradient.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import net.kyori.adventure.text.format.TextColor;
+import net.kyori.adventure.util.HSVLike;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A Gradient in the HSV color space.
+ *
+ * @since 4.15.0
+ */
+public class HSVGradient extends HueGradient {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/text/format/gradient/HSVGradient.java:35]: All API types should be `interface` - implementations should be package-private.

File: api/src/main/java/net/kyori/adventure/text/format/gradient/HueGradient.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+/**
+ * A HueGradient is a gradient whose {@code ColorSpace} involves some form of hue coordinate, which needs
+ * to be specially interpolated.
+ *
+ * @see <a href="https://www.w3.org/TR/css-color-4/#hue-interpolation">W3C CSS Color Level 4 spec on hue interpolation</a>
+ *
+ * @since 4.15.0
+ */
+public abstract class HueGradient implements Gradient {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/text/format/gradient/HueGradient.java:36]: All API types should be `interface` - implementations should be package-private.

File: api/src/main/java/net/kyori/adventure/text/format/gradient/OKLabGradient.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+import net.kyori.adventure.util.LABLike;
+import net.kyori.adventure.util.RGBLike;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A gradient in the OkLab Colorspace.
+ *
+ * @since 4.15.0
+ */
+public class OKLabGradient implements Gradient {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/text/format/gradient/OKLabGradient.java:36]: All API types should be `interface` - implementations should be package-private.

File: api/src/main/java/net/kyori/adventure/text/format/gradient/RGBGradient.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+import net.kyori.adventure.text.format.TextColor;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A gradient in the RGB Colorspace.
+ *
+ * @since 4.15.0
+ */
+public class RGBGradient implements Gradient {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/text/format/gradient/RGBGradient.java:34]: All API types should be `interface` - implementations should be package-private.

File: api/src/main/java/net/kyori/adventure/util/LabLikeImpl.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
+package net.kyori.adventure.util;
+
+import net.kyori.adventure.internal.Internals;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Objects;
+
+public class LabLikeImpl implements LABLike {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/util/LabLikeImpl.java:8]: Implementations should be package-private.

**Review by kashike — CHANGES_REQUESTED** [review]

**Comment by thiccaxe** [comment]
decided to rework this in https://github.com/thiccaxe/adventure-gradient

Resolves the main concern in

> Idk if it's just me, but it seems weird to have interpolation methods on the gradient? I don't really have an better API design but something about that feels off to me. Like, they are just util methods and can be used for other stuff, its a bit weird to tie them to the concept of a gradient?

Now interpolation is kept separate from the gradient business. No need for extra implementations per-gradient type.

**Pull request closed without merging**